bugwarrior/services/gmail.py

```diff
@@ -0,0 +1,178 @@
+import httplib2
+import os
+import email
+
+import googleapiclient.discovery
+import oauth2client.client
+import oauth2client.tools
+import oauth2client.file
+
+from bugwarrior.config import die
+from bugwarrior.services import IssueService, Issue
+
+import logging
+log = logging.getLogger(__name__)
+
+class GmailIssue(Issue):
+    THREAD_ID = 'gmailthreadid'
+    SUBJECT = 'gmailsubject'
+    URL = 'gmailurl'
+    LAST_SENDER = 'gmaillastsender'
+    LAST_SENDER_ADDR = 'gmaillastsenderaddr'
+    SNIPPET = 'gmailsnippet'
+
+    UNIQUE_KEY = (THREAD_ID,)
+    UDAS = {
+        THREAD_ID: {
+            'type': 'string',
+            'label': 'GMail Thread Id',
+        },
+        SUBJECT: {
+            'type': 'string',
+            'label': 'GMail Subject',
+        },
+        URL: {
+            'type': 'string',
+            'label': 'GMail URL',
+        },
+        LAST_SENDER: {
+            'type': 'string',
+            'label': 'GMail last sender name',
+        },
+        LAST_SENDER_ADDR: {
+            'type': 'string',
+            'label': 'GMail last sender address',
+        },
+        SNIPPET: {
+            'type': 'string',
+            'label': 'GMail snippet',
+        }
+    }
+    EXCLUDE_LABELS = [
+        'IMPORTANT',
+        'CATEGORY_PERSONAL',
+        'CATEGORY_PROMOTIONS',
+        'CATEGORY_UPDATES',
+        'CATEGORY_FORUMS',
+        'SENT']
+
+    def to_taskwarrior(self):
+        return {
+            'tags': [label
+                for label in self.extra['labels']
+                if label not in self.EXCLUDE_LABELS],
+            'priority': self.origin['default_priority'],
+            self.THREAD_ID: self.record['id'],
+            self.SUBJECT: self.extra['subject'],
+            self.URL: self.extra['url'],
+            self.LAST_SENDER: self.extra['last_sender_name'],
+            self.LAST_SENDER_ADDR: self.extra['last_sender_address'],
+            self.SNIPPET: self.extra['snippet'],
+        }
+
+    def get_default_description(self):
+        return self.build_default_description(
+            title=self.extra['subject'],
+            url=self.get_processed_url(self.extra['url']),
+            number=self.record['id'],
+            cls='issue',
+        )
+
+class GmailService(IssueService):
+    APPLICATION_NAME = 'Bugwarrior Gmail Service'
+    SCOPES = 'https://www.googleapis.com/auth/gmail.readonly'
+
+    ISSUE_CLASS = GmailIssue
+    CONFIG_PREFIX = 'gmail'
+
+    def __init__(self, *args, **kw):
+        super(GmailService, self).__init__(*args, **kw)
+
+        self.query = self.config.get('query', 'label:Starred')
+        self.login_name = self.config.get('login_name', 'me')
+        self.client_secret_path = self.get_config_path('client_secret_path')
+        self.credentials_path = self.get_config_path('credentials_path')
```

Comment:
Would it make sense to have default paths for these? In fact, are settings really needed? I have an especially hard time seeing the use case for `credentials_path` since the user never has any reason to interact with the file.